adapters/syslog/syslog_test.go

```diff
@@ -1,22 +1,210 @@
 package syslog
 
 import (
+	"fmt"
+	"net"
 	"os"
 	"strconv"
+	"sync"
 	"testing"
+	"text/template"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/gliderlabs/logspout/router"
+
+	_ "github.com/gliderlabs/logspout/transports/tcp"
+	_ "github.com/gliderlabs/logspout/transports/tls"
+	_ "github.com/gliderlabs/logspout/transports/udp"
 )
 
-func TestSyslogRetryCount(t *testing.T) {
-	setRetryCount()
-	if retryCount != defaultRetryCount {
-		t.Errorf("expected %v got %v", defaultRetryCount, retryCount)
+const maxMsgCount = 10
+
+var (
+	container = &docker.Container{
+		ID:   "8dfafdbc3a40",
+		Name: "0michaelshobbs",
+		Config: &docker.Config{
+			Hostname: "8dfafdbc3a40",
+		},
+	}
+	testPriority  = "{{.Priority}}"
+	testTimestamp = "{{.Timestamp}}"
+	testHostname  = "{{.Container.Config.Hostname}}"
+	testTag       = "{{.ContainerName}}"
+	testPid       = "{{.Container.State.Pid}}"
+	testData      = getopt("SYSLOG_DATA", "{{.Data}}")
+	testTmplStr   = fmt.Sprintf("<%s>%s %s %s[%s]: %s\n",
+		testPriority, testTimestamp, testHostname, testTag, testPid, testData)
+)
+
+type localTCPServer struct {
+	lnmu sync.RWMutex
+	net.Listener
+}
+
+func (ls *localTCPServer) teardown() error {
+	ls.lnmu.Lock()
+	if ls.Listener != nil {
+		ls.Listener.Close()
+		ls.Listener = nil
 	}
+	ls.lnmu.Unlock()
+	return nil
+}
+
+func newLocalTCPServer() (*localTCPServer, error) {
+	ln, err := newLocalListener()
+	if err != nil {
+		return nil, err
+	}
+	return &localTCPServer{Listener: ln}, nil
+}
+
+func newLocalListener() (net.Listener, error) {
+	ln, err := net.Listen("tcp4", "127.0.0.1:0")
+	if err != nil {
+		return nil, err
+	}
+	return ln, nil
+}
+
+func readConn(c net.Conn, ch chan []byte) error {
+	b := make([]byte, 256)
+	_, err := c.Read(b)
+	if err != nil {
+		return err
+	}
+	ch <- b
+	return nil
+}
 
+func TestSyslogRetryCount(t *testing.T) {
 	newRetryCount := uint(20)
 	os.Setenv("RETRY_COUNT", strconv.Itoa(int(newRetryCount)))
-	defer os.Unsetenv("RETRY_COUNT")
 	setRetryCount()
 	if retryCount != newRetryCount {
 		t.Errorf("expected %v got %v", newRetryCount, retryCount)
 	}
+
+	os.Unsetenv("RETRY_COUNT")
+	setRetryCount()
+	if retryCount != defaultRetryCount {
+		t.Errorf("expected %v got %v", defaultRetryCount, retryCount)
+	}
+}
+
+func TestSyslogReconnectOnClose(t *testing.T) {
```

Comment:
This test is pretty intense.
I'd recommend breaking it down a little. Maybe move the `go func()` to named functions so they're just `go someReallyClearFuncName()`
Would make things a little more readable.